src/core/jit/exec.rs

```diff
@@ -117,7 +117,11 @@ where
                     if let Some(array) = value.as_array() {
                         join_all(field.nested_iter().map(|field| {
                             join_all(array.iter().enumerate().map(|(index, value)| {
-                                let ctx = ctx.with_value_and_field(value, field); // Output::JsonArray::Value
+                                let mut ctx = ctx.with_value_and_field(value, field); // Output::JsonArray::Value
+                                if !field.is_scalar {
+                                    let new_value = value.get_key(&field.name).unwrap_or(value);
+                                    ctx = ctx.with_value_and_field(new_value, field);
+                                }
```

Comment:
Ok, I think I get it.

We need to update value on every level of nesting since the resolver may require it and the value should be actually the direct parent value for the place where current IR is defined, not some ancestor value that was resolved by some ancestor IR.

But, couple of moments:
- is_scalar is redundant. Scalars should not have children fields at all i.e. `field.nested_iter()` will return empty iterator
- this logic for updating the value should be executed for any other cases when we're calling `field.nested_iter()` and it's not specific to arrays only. And, please note, that for your newly added code for no IR case the actual parent value you want to use is the value stored in `ctx` since it should be populated by the previous `execution` call